[HLSL] Rewrite semantics parsing

Keenuts opened the pull request:
This is the first PR to implement the semantics proposal: https://github.com/llvm/wg-hlsl/blob/main/proposals/0031-semantics.md

This PR focuses on the changes required to handle user semantics, but tried to be almost NFC.
What changes is the error messages as the semantics case is not kept when reporting error messages.

You might notice the SV_GroupIndex semantic is not properly validated as are others. This is an existing behavior that we'll need to fix, but wanted to keep this separated from this rewrite to stay as-close as an NFC as possible.

The next PR will add support on the different kinds of I/O we can have using semantics (input, inout param, structs).

Keenuts committed:
[HLSL] Rewrite semantics parsing

This is the first PR to implement the semantics proposal:
https://github.com/llvm/wg-hlsl/blob/main/proposals/0031-semantics.md

This PR focuses on the changes required to handle user semantics, but
tried to be almost NFC.
What changes is the error messages as the semantics case is not kept
when reporting error messages.

You might notice the SV_GroupIndex semantic is not properly validated as
are others. This is an existing behavior that we'll need to fix, but
wanted to keep this separated from this rewrite to stay as-close as an
NFC as possible.

The next PR will add support on the different kinds of I/O we can have
using semantics (input, inout param, structs).

Keenuts commented:
@llvm-beanz anything blocking this review?

s-perron reviewed (commented):
My only concern is that the parser does not allow digits in the middle of the semantic name. I believe that is possible for user semantics. It works for DXC. Otherwise LGTM.

shafik commented on clang/include/clang/AST/Attr.h:
```diff
@@ -232,6 +232,41 @@ class HLSLAnnotationAttr : public InheritableAttr {
+      : HLSLAnnotationAttr(Context, CommonInfo, AK, IsLateParsed,
+                           InheritEvenIfAlreadyPresent) {
+    this->SemanticIndexable = SemanticIndexable;
+    this->SemanticIndex = 0;
```
If these are sensible default values then why not use in class member initializers? These are safer under refactoring over time.

shafik commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -393,6 +393,14 @@ def warn_hlsl_langstd_minimal :
 
+def err_hlsl_semantic_index_overlap : Error<"semantic index overlap %0">;
+
+def err_hlsl_semantic_missing : Error<"semantic annotations must be present "
```
So I see new diagnostic but no tests that seem to exercise them? This does for every diagnostic not tested.

Keenuts replied:
Added a test for the missing semantic, and removed the 2 others which should be added in a later PR.

shafik commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -5188,6 +5188,14 @@ class Parser : public CodeCompletionHandler {
   }
 
+  struct ParsedSemantic {
+    StringRef Name;
```
Based on the implementation of `ParseHLSLSemantic` all these fields have default values, you should use in class member initializers and just return the defaulted values instead of manually setting the same values three times. This will be very error prone under refactoring and likely won't remain consistent.

shafik reviewed (commented):
I may come back to this.

Keenuts commented on clang/lib/CodeGen/CGHLSLRuntime.h:
```diff
@@ -172,6 +193,9 @@ class CGHLSLRuntime {
 
   llvm::DenseMap<const clang::RecordType *, llvm::TargetExtType *> LayoutTypes;
+
+  std::unordered_set<std::string> ActiveInputSemantics;
```
Ah right, I should move this to one of the children commit (and change to StringSet), thanks!

farzonl commented:
> @farzonl anything blocking?

I would prefer if @llvm-beanz  or @tex3d did the sign off for this pr from our side.

Keenuts commented:
@llvm-beanz friendly ping

llvm-beanz reviewed (commented):
My only big concern is around the forced capitalization of semantics in the diagnostics.

I also left some notes about follow-up changes we should consider.

Keenuts committed:
preserve semantic name case for diagnostics

Keenuts committed:
formatting

Keenuts committed:
replace manual loop with find_last_not_of

Keenuts commented on clang/lib/Parse/ParseHLSL.cpp:
```diff
@@ -118,6 +118,47 @@ static void fixSeparateAttrArgAndNumber(StringRef ArgStr, SourceLocation ArgLoc,
+
+  assert(Identifier.size() > 0);
+  unsigned I = Identifier.size();
+  for (; I > 0 && isDigit(Identifier[I - 1]); --I)
```
Replaced this loop with `find_last_not_of` which in our case is a good replacement

Keenuts commented:
Thanks for the reviews! Applied the last small bits

Keenuts committed:
fix unused variable

The new system doesn't require the casted value, so we had a unused
variable warning.

mgorny commented:
I'm guessing this change broke the build for me:

```
FAILED: [code=1] docs/AttributeReference.rst /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/docs/AttributeReference
.rst 
cd /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/docs && /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/cla
ng-.arm64/bin/clang-tblgen -gen-attr-docs -I/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs -I/var/tmp/portage/llvm-core/c
lang-22.0.0.9999/work/clang/include -I/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/include -I/usr/lib/llvm/22/inc
lude -I/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs/../include/clang/Basic /var/tmp/portage/llvm-core/clang-22.0.0.9999
/work/clang/docs/../include/clang/Basic/Attr.td --write-if-changed -o AttributeReference.rst -d AttributeReference.rst.d && /usr/bin/cm
ake -E cmake_transform_depfile Ninja gccdepfile /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang /var/tmp/portage/llvm-core/clan
g-22.0.0.9999/work/clang/docs /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64 /var/tmp/portage/llvm-core/clang-22.0.
0.9999/work/x/y/clang-.arm64/docs /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/docs/AttributeReference.rst.d /var
/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/CMakeFiles/d/63214aa3df9ce54c496015feb0fb3942e9cf73781c78cb807e54fac358a
06d47.d
/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs/../include/clang/Basic/Attr.td:4986:5: error: Attribute has no supported spellings; cannot be documented
def HLSLSV_DispatchThreadID : HLSLSemanticAttr</* Indexable= */ 0> {
    ^
```

Keenuts commented:
> I'm guessing this change broke the build for me:
> 
> ```
> FAILED: [code=1] docs/AttributeReference.rst /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/docs/AttributeReference.rst 
> cd /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/docs && /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/bin/clang-tblgen -gen-attr-docs -I/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs -I/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/include -I/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/include -I/usr/lib/llvm/22/include -I/var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs/../include/clang/Basic /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs/../include/clang/Basic/Attr.td --write-if-changed -o AttributeReference.rst -d AttributeReference.rst.d && /usr/bin/cmake -E cmake_transform_depfile Ninja gccdepfile /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64 /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/docs /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/docs/AttributeReference.rst.d /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/x/y/clang-.arm64/CMakeFiles/d/63214aa3df9ce54c496015feb0fb3942e9cf73781c78cb807e54fac358a06d47.d
> /var/tmp/portage/llvm-core/clang-22.0.0.9999/work/clang/docs/../include/clang/Basic/Attr.td:4986:5: error: Attribute has no supported spellings; cannot be documented
> def HLSLSV_DispatchThreadID : HLSLSemanticAttr</* Indexable= */ 0> {
>     ^
> ```

Interesting, what cmake flags do you use? (seems like the CI doesn't build with those, so maybe I can repro on my end?)

mgorny commented:
Thanks!

mgorny commented:
And sorry, I won't be able to take a closer look at it today, as the devboxes are currently busy building new kernels.

Keenuts commented:
I was able to reproduce the build issue. Turns out we'd benefit from a different documentation generation for those. Will re-land once #157841 is merged.
